fix: Markdown link syntax (#56)
cloudnativegeo/cloud-optimized-geospatial-formats-guide: cloud-optimized-geotiffs/cogs-details.qmd

---
title: "Make better GeoTIFFs"
---

> # Make better GeoTIFFs 

Reviewer: All the rest of our sidebar titles and h1 headers are nouns instead of verbs; this reads a little weird to me.

Maybe `Advanced COG Details` or `Advanced COG Recommendations`? Each of our pages could have `Advanced *`. It also seems nice to have the title in the sidebar be the same or roughly the same as the URL fragment

In the [COG Intro](intro.qmd) you can see what makes a COG different from non-optimized GeoTIFFs. The rest of this page details additional GeoTIFF information that can be relevant to making your files as useful as possible but not a COG requirement.

## Data Type

> **Recommendation** The smallest possible data type, that still represents the data appropriately, should be used. It is not generally recommended to shift data from float to integer by multiplying, a space saving technique, as end users then need to undo this step to use the data. Data compression is preferred, see also [Compression](#compression).

Reviewer: ```suggestion
**Recommendation** The smallest possible data type that still represents the data appropriately should be used. It is not generally recommended to shift or quantize data from float to integer by multiplying, a space saving technique, as end users then need to undo this step to use the data. Data compression is preferred, see also [Compression](#compression).
```

> GeoTIFF format supports many data types. The key is that all bands must be of the same data type. Unlike some other formats you can not mix and match integers (whole numbers) and floats (decimal numbers) in the same file. If you have this use case consider splitting files by data type and using a catalog like STAC to keep track of them, or look at other formats like [Zarr](../zarr/intro.qmd).

Reviewer: ```suggestion
The GeoTIFF format supports many data types. The key is that all bands must be of the same data type. Unlike some other formats, you cannot mix and match integer (whole number) and float (decimal number) data types in the same file. If you have this use case consider splitting files by data type and using a catalog like STAC to keep track of them, or look at other formats like [Zarr](../zarr/intro.qmd).
```

Scenario: If the COG is intended only for visualization, conversion to 3 band byte will improve performance. 

> GDAL supported Data Types [list](https://gdal.org/drivers/raster/gtiff.html#gtiff-geotiff-file-format)

## Compression (aka File Size)

> The biggest benefit to compression is on the storage side. It’s always recommended to use a lossless compression method. **Deflate** or **LZW** are the most recommended compression algorithms, there are some choices that depend on the data type and distribution, and if the goal is maximum compression or not. Maximum compression does result in some performance loss.

Reviewer: ```suggestion
The biggest benefit to compression is on the storage side. It’s always recommended to use a lossless compression method unless the COG is intended to be used only for visualization. **Deflate** or **LZW** are the recommended compression algorithms for general use. Some other compression algorithms may be preferred depending on the data type and distribution, and if the goal is maximum compression or not. Maximum compression does result in some performance loss.
```

## No Data

> Setting a no data value makes it clear to users and visualization tools what pixels are not actually data. For visualization this allows these pixels to be easily hidden (transparent). Historically many values have been used, 0, -9999, etc… The key is to make sure the GDAL flag for no data is set. It is also suggested that the smallest negative value be used instead of a random value. For byte and unsigned integers/floats this will be 0, if 0 has meaning in your data use a different value (like the max possible value). Having the right nodata flag set is important for overview generation.

Reviewer: Can/should `NaN` be a no data value for float data types?

Author: Oh this is hotly debated. It is acceptable, except NASA DAACs prefer not. I'll try to dig up why.

## Projection

Read performance can be greatly impacted by the choice of projection and the particular applications used for dynamic tile serving. Using a known CRS defined in the PROJ database (typically EPSG code) is preferred over custom projections. Load times can be 5-20 times greater when using a custom projection. Whenever applying projections make sure to use WKT2 representation. If using a database of known projections, i.e. EPSG codes, this should be fine, there are known issues around manually setting proj-strings.

## What we don’t know (areas of research)

* The optimum size of data at which splitting across files improves performance as a multi-file dataset instead of a single file.
* When to recommend particular internal tile sizes

> * Compression impacts on http transfer rates.

Reviewer: What is this supposed to mean? Does this mean "full time to load and parse an image in a browser"? Or "Compression speed for streaming compression"? Except maybe for streaming compression/decompression on a server, the only impact on http transfer rates would be the size of the file...?

> * Support for COG creation in all common Geospatial tools varies.

Reviewer: ```suggestion
* Support for COG creation in all common geospatial tools varies.
```
